// react-chrome-redux/lib/store/Store.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assignIn = require('lodash/assignIn');

var _assignIn2 = _interopRequireDefault(_assignIn);

var _constants = require('../constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Store = function () {
  function Store(_ref) {
    var _this = this;

    var portName = _ref.portName;
    var _ref$state = _ref.state;
    var state = _ref$state === undefined ? {} : _ref$state;

    _classCallCheck(this, Store);

    this.port = chrome.runtime.connect({ name: portName });
    this.listeners = [];
    this.state = state;

    this.port.onMessage.addListener(function (message) {
      if (message.type === _constants.STATE_TYPE) {
        _this.replaceState(message.payload);
      }
    });

    this.dispatch = this.dispatch.bind(this);
  }

  _createClass(Store, [{
    key: 'subscribe',
    value: function subscribe(listener) {
      var _this2 = this;

      this.listeners.push(listener);

      return function () {
        _this2.listeners = _this2.listeners.filter(function (l) {
          return l !== listener;
        });
      };
    }
  }, {
    key: 'replaceState',
    value: function replaceState(state) {
      this.state = state;

      this.listeners.forEach(function (l) {
        return l();
      });
    }
  }, {
    key: 'getState',
    value: function getState() {
      return this.state;
    }
  }, {
    key: 'dispatch',
    value: function dispatch(data) {
      return new Promise(function (resolve, reject) {
        chrome.runtime.sendMessage({
          type: _constants.DISPATCH_TYPE,
          payload: data
        }, function (_ref2) {
          var error = _ref2.error;
          var value = _ref2.value;

          if (error) {
            reject((0, _assignIn2.default)(new Error(), error));
          } else {
            resolve(value.payload);
          }
        });
      });
    }
  }]);

  return Store;
}();

exports.default = Store;
